packages/shared/src/contexts/UITaskManager.tsx

```diff
@@ -0,0 +1,99 @@
+import { defer, DeferTuple } from '@dimensiondev/kit'
+import { EMPTY_LIST } from '@masknet/shared-base'
+import { createContext, createElement, FC, FunctionComponent, PropsWithChildren, useMemo, useState } from 'react'
+import type { InjectedDialogProps } from './components'
+
+type ExtendPromise<T> = Promise<T> & {
```

Comment:
Can you explain the use case of this?

Reply:
Intend to provide a way to close a dialog programmably.
There is no use case right now indeed.

Reply:
Is it possible to use things like `{promise: Promise<T>, close?: () => void}`?

Reply:
> Is it possible to use things like `{promise: Promise<T>, close?: () => void}`?

Then the common usage would be wordy

```ts
const openDialog = useOpenDialog()
await openDialog(options).promise
```

----

> Intend to provide a way to close a dialog programmably.

Use case for `close()`

```ts
const openDialog = useOpenDialog()
const promise = openDialog(options)
const timer = setTimeout(promise.close, 1e5)
await promise
clearTimeout(timer)
```

Reply:
OK, this is interesting. And where are you attaching the `close` method?

Reply:
```ts
                    const extendedPromise = Object.assign(promise, {
                        close() {
                            resolve(null)
                            return promise
                        },
                    })
                    const newTask: Task = { id, promise, resolve, reject, taskOptions: options }
                    setTasks((list) => [...list, newTask])
                    promise.then(() => {
                        removeTask(id)
                    })
                    return extendedPromise
```